dartagnan/src/test/java/com/dat3m/dartagnan/witness/BuildWitnessTest.java

```diff
@@ -0,0 +1,65 @@
+package com.dat3m.dartagnan.witness;
+
+import com.dat3m.dartagnan.parsers.cat.ParserCat;
+import com.dat3m.dartagnan.parsers.program.ProgramParser;
+import com.dat3m.dartagnan.program.Program;
+import com.dat3m.dartagnan.utils.ResourceHelper;
+import com.dat3m.dartagnan.utils.Result;
+import com.dat3m.dartagnan.utils.Settings;
+import com.dat3m.dartagnan.utils.TestHelper;
+import com.dat3m.dartagnan.utils.options.DartagnanOptions;
+import com.dat3m.dartagnan.verification.VerificationTask;
+import com.dat3m.dartagnan.wmm.Wmm;
+import com.dat3m.dartagnan.wmm.utils.Arch;
+
+import org.junit.Test;
+import org.sosy_lab.java_smt.api.BooleanFormula;
+import org.sosy_lab.java_smt.api.BooleanFormulaManager;
+import org.sosy_lab.java_smt.api.ProverEnvironment;
+import org.sosy_lab.java_smt.api.SolverContext;
+import org.sosy_lab.java_smt.api.SolverContext.ProverOptions;
+
+import static com.dat3m.dartagnan.analysis.Base.runAnalysisIncrementalSolver;
+
+import java.io.File;
+
+public class BuildWitnessTest {
+
+    @Test
+    public void BuildWriteEncode() throws Exception {
+    	Program p = new ProgramParser().parse(new File(ResourceHelper.TEST_RESOURCE_PATH + "witness/lazy01-O0.bpl"));
+    	Wmm wmm = new ParserCat().parse(new File(ResourceHelper.CAT_RESOURCE_PATH + "cat/svcomp.cat"));
+    	VerificationTask task = new VerificationTask(p, wmm, Arch.NONE, new Settings(0, 0));
+
+    	try (SolverContext ctx = TestHelper.createContext();
+    			ProverEnvironment prover = ctx.newProverEnvironment(ProverOptions.GENERATE_MODELS))
+    	{
+    		Result res = runAnalysisIncrementalSolver(ctx, prover, task);
+    		// This code mostly tests the parsing of witness options thus the "ignore"
+    		// The only parts that are needed to build the graph are the unrolling bound ([4] and [5])
+    		// and the file to compute the SHA256 value ([6] and [7]).
+    	    String[] sOptions = new String[8];
+    	    sOptions[0] = "-i";
+    	    sOptions[1] = "ignore.bpl";
+    	    sOptions[2] = "-cat";
+    	    sOptions[3] = "ignore";
+    	    sOptions[4] = "-unroll";
+    	    sOptions[5] = "1";
+    	    sOptions[6] = "-create_witness";
+    	    sOptions[7] = ResourceHelper.TEST_RESOURCE_PATH + "witness/lazy01-O0.bpl";
+    	    DartagnanOptions option = new DartagnanOptions();
+    	    option.parse(sOptions);
+    	    
+    		WitnessGraph graph = new WitnessBuilder(p, ctx, prover, res).buildGraph(option);
+    		// Write to file
+    		graph.write();
+    		assert(new File(System.getenv("DAT3M_HOME") + "/output/witness.graphml").exists());
```

Comment:
I think the file never gets deleted so this test might be meaningless. Maybe we can leave this part out for now.

Reply:
Keep in mind that each instance of the CI runs in a different "container" (not sure if this is the correct term) meaning that the file from the previous run of the CI does not exists and thus the test still checks that the file was correctly created in this run.

Reply:
Ok, for the CI it makes sense then. It won't work locally though. Maybe add a comment that this works for the CI only?

Reply:
I can also delete the file create by this tests ... this will not guarantee 100% that the test is not trivial locally because you can still have a witness file even if the test failed because e.g. you used the svcomp runner. Another way to reduce the chances of this happening is to use the program name for the witness file instead of just `witness.graphml` (I need to double check this would not cause problems for svcomp but I don't think so).

Reply:
Different names is probably the best idea. But even if you just delete the file at the end (which you probably should either way), the test would only be trivial for the first run since it will delete whatever file is there, making sure that the second run is an actual test (unless you create a new file inbetween each two tests, but that seems unlikely).

Reply:
The new version even checks that the file does not exist before writing the witness. I also modified the bpl filename to avoid the situation where the user has a witness file due to using the svcomp runner with the same task we use for the unit test.